rfcs/0000-terraform-workflow-improvements.md

```diff
@@ -0,0 +1,826 @@
+
+- Prod deployment requires prerequisite deployment of IAM resource module
+- Engineers do not require IAM permissions to deploy changes to prod
+
+#### Risks
+
+- Onerous process to deploy IAM modules separately from application modules
+- Module testing requires creating IAM resources separately
+- Engineers can still define IAM resources anywhere, though they'll lack the permissions to apply them
+
+# Prior art
+[prior-art]: #prior-art
+
+<!--
+Discuss prior art, both the good and the bad, in relation to this proposal. A few examples of what
+this can include are:
+
+- Can we learn something about this proposal from other projects in CTD?
+- Can we learn something about this proposal from other transit agencies?
+- Can we learn something about this proposal from past experiences in other jobs or projects?
+
+This section is intended to encourage you as an author to think about the lessons from other places,
+and provide readers of your RFC with a fuller picture.
+
+If there is no prior art, that is fine.
+
+Note that while precedent is some motivation, it does not on its own motivate an RFC.
+-->
+
+Workflow automation for Terraform is entirely new territory for the authors of this document. We've done a fair amount of research and factfinding to formulate a plan that we think brings CTD's usage of Terraform more in line with industry best practices, and takes advantage of products that take some of the work off our shoulders. That said, it's entirely possible there's something we missed, and we're open to recommendations of solutions, software or patterns from anyone who cares to weigh in.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+<!--
+- What parts of the design do you expect to resolve through the RFC process before this gets merged?
+- What parts of the design do you expect to resolve through the implementation of this feature
+  before stabilization?
+- What related issues do you consider out of scope for this RFC that could be addressed in the
+  future independently of the solution that comes out of this RFC?
+-->
+
+One major unresolved question is the timeline for implementing all the improvements laid out in this RFC. As noted in the [Drawbacks][drawbacks] section above, it's a lot of work. Once this RFC is approved we'll need to scope out the work and divide it accordingly between the infrastructure and engineering teams.
+
+Other questions for the reader:
+
+- Is the proposed [root module naming convention][root-module-naming-convention] sensible and sufficiently comprehensive?
+- Does the [isolated module testing workflow][reference-module-testing] make sense? Does it seem easy enough to carry out?
+- How aggressive should we be about [standardizing on infrastructure component modules][guide-standardization] for ECS and RDS across all applications/teams?
+- How aggressive should we be about [versioning and publishing application modules][guide-module-versioning] for open-source applications?
```

Comment:
Not at all. If anybody else wants to use an open-source application, they're probably going to have a completely different infrastructure and deployment setup than us. It's perhaps most useful as a reference of what a working deployment is like, which env vars are filled in which ways, but I wouldn't expect anyone to take a module for an open source repo and deploy it as is.